svelte: Add infinity scroll to commit page and simplify history panel implementation

fkling opened the pull request:
This commit adds infinity scroll behavior to the commit page, building on top of [Apollo's pagination features.](https://www.apollographql.com/docs/react/pagination/overview). This has the advantage of not having to introduce custom logic in the component for handling pagination, which makes the components simpler. Now they only need to work with a data list and a function to fetch more data. **Update:** Using the Apollo cache for this is not as straightforward as it seemed. See [my follow up comment](https://github.com/sourcegraph/sourcegraph/pull/59639#issuecomment-1897416068)

Additionally [SvelteKit's snapshot feature](https://kit.svelte.dev/docs/snapshots) is used to track the position of the scrollbar and the expanded/collapsed state of individual file diffs.

A drawback is that the field policy code is not very strongly typed.

I applied the same technique to the history panel which simplifies its implementation a lot.


## Test plan

- Scroll on commit page. A loading indicator is shown when more diffs are being loaded and replaced by the new data.
- Scroll history panel, same as above. Additionally, when switching between files the previous list of commits should be removed and a loading indicator should be shown instead.

fkling commented on client/web-sveltekit/src/lib/repo/HistoryPanel.svelte:
```diff
@@ -48,59 +44,64 @@
 
     $: selectedRev = $page.url?.searchParams.get('rev')
     $: clearURL = getClearURL()
-    $: canShowInlineDiff = $page.route.id?.includes('/blob/')
```
Changing this makes the component's story work again. `$page` is not available in Storybook.

camdencheek commented on client/web-sveltekit/src/lib/Scroller.svelte:
```diff
@@ -14,7 +14,18 @@
+                    window.requestAnimationFrame(syncScroll)
+                }
+            }
+        })
```
Something I noticed while playing around: scrolling to the bottom of the history starts scrolling the blob view while loading the next page.

https://github.com/sourcegraph/sourcegraph/assets/12631702/fcaf9197-5fcb-45a2-84f0-78e911f2bdf1

fkling replied:
I fixed that.

camdencheek reviewed (commented):
Trying it out locally, this seems to have broken pagination on the repository commits page. I don't see any scroll-activated pagination on this page, and the "Next" button doesn't work

https://github.com/sourcegraph/sourcegraph/assets/12631702/dbac3ba5-90ce-4a0f-b1fa-2c359e79ad55

camdencheek commented:
The inline diff is really nice! One thing I noticed is there is no way to view the full commit message in the history panel -- only the commit summary

camdencheek commented:
The full-page scroll into the simple UI mode feels really weird with a fixed-size history panel. Just makes it feel like the page is scrolling behind the history panel. 

https://github.com/sourcegraph/sourcegraph/assets/12631702/3e99b9ba-c841-46f5-a7cc-ec775cc3e4af

fkling commented on client/web-sveltekit/src/lib/graphql/apollo.ts:
```diff
@@ -52,9 +53,51 @@ const customFetch: HttpOptions['fetch'] = (uri, options) => fetch(uri, options).
+            }
+            const nodes = [...existing.nodes]
+            const offset = args[cursorName] ? +args[cursorName] : 0
+            for (let i = 0; i < incoming.nodes.length; ++i) {
+                nodes[offset + i] = incoming.nodes[i]
+            }
```
I've been following the suggestions in Apollos docs. This ensures that the responses from "out of band" requests are properly ordered. E.g. imagine a request with `{after: '40'}` is executed before `{after: '20'}`. If we just concatenate the nodes would be in the wrong order.

fkling commented on client/web-sveltekit/src/lib/graphql/apollo.ts:
```diff
@@ -52,9 +53,51 @@ const customFetch: HttpOptions['fetch'] = (uri, options) => fetch(uri, options).
+                return incoming
+            }
+            const nodes = [...existing.nodes]
+            const offset = args[cursorName] ? +args[cursorName] : 0
```
Yeah, I understand that. It's hacky right now.

I guess what's missing from our responses then is a way to get the cursor of a specific node in the list. E.g. in the case of commits, if the cursor was the ID of the commit, then I could just find the appropriate index by looking through the existing list of nodes for the commit with the specific ID. 

Maybe this helps with understanding what I'm talking about: https://www.apollographql.com/docs/react/pagination/cursor-based

camdencheek replied:
I don't think that's really possible with how we designed our pagination APIs. For example, a cursor can be a user (IDs aren't guaranteed to be sequential), commit hash (next page returns commits older than that hash), or the combination of repo stars and repo ID (which is definitely not representible as a number). Additionally, many of our GraphQL IDs are base64-encoded integers rather than actual integers. 

All that to say, I don't think this hack will get us very far. We should match our client-side pagination with the API provided by the backend, which is an _opaque_ cursor that is passed back to the backend to get the next page. We cannot generally calculate a cursor for an arbitrary item. 

I don't _think_ it's a big deal that we can't append at arbitrary locations -- I don't expect us to have multiple "next page" requests happening concurrently.

fkling replied:
I think you misunderstood me. I'm not saying that the cursor needs to be a number that it needs to be computable on the client side. I'm saying there should be a way for the client to request the cursor from the server. E.g. in Relay you'd do this:

```
ancestors(...) {
  edges {
    cursor
    node {
      ...
    }
  }
  pageInfo {...}
}
```

In this case `cursor` can be completely opaque. This way the client can manage 

> I don't think it's a big deal that we can't append at arbitrary locations -- I don't expect us to have multiple "next page" requests happening concurrently, so just concatenating the new results seems fine to me

That will fail if you clicked `Next`, refresh the page and then press `Previous`. 

FWIW, I would prefer a different solution, but that seems to be the only way to this right now for these specific queries.

I'll take some more time to think of a solution with less assumptions.

The nice thing about merging different pagination results is that we don't need any additional logic to manage the state of an infinitely scrollable list. If you navigate away from the list and go back to it, the query will return all results that have been fetched so far, not just the ones that are requested via the parameters.

Of course an alternative approach would be to not have infinity scroll at all and always use proper pagination, but I don't know if that's suitable for every situation.

fkling replied:
Not an issue anymore since I switch to a simpler implementation that doesn't need to work with cursor.

fkling commented:
@camdencheek Thank you for the thorough tests!

> Trying it out locally, this seems to have broken pagination on the repository commits page. I don't see any scroll-activated pagination on this page, and the "Next" button doesn't work

Yeah, my changes to merging the lists broke that. I added a way that allows proper pagination to work by aliasing the field. No idea if that's a good idea :D 
I also added an integration test for the page to test the navigation, which uncovered two other issues.

> Something I noticed while playing around: scrolling to the bottom of the history starts scrolling the blob view while loading the next page.

I fixed that by adding [`overscroll-behavior-y: contain`](https://developer.mozilla.org/en-US/docs/Web/CSS/overscroll-behavior). I should add that in other places too.

> The inline diff is really nice! One thing I noticed is there is no way to view the full commit message in the history panel -- only the commit summary

There hasn't been any kind of review for the panel yet. I just implemented something that made sense to me at the time.

> The full-page scroll into the simple UI mode feels really weird with a fixed-size history panel. Just makes it feel like the page is scrolling behind the history panel.

Similarly, full-page scroll had been an idea and I wanted to see how easy/difficult it would be to add. Eventually the history should be resizable too (or maybe we go with a completely different solution than a panel 🤷🏻 ).

fkling commented:
> nit: the offset of the hunk header looks weird to me because it's a visual delimiter between hunks, but it's indented. I think we should either left-align it or visually emphasize it more strongly


FWIW the whole inline diff view is also just put together with minimal effort. No design review went into this yet.

camdencheek commented:
>  put together with minimal effort 

It's impressively nice for "minimal" -- had me fooled!

fkling commented:
Alright, the commit**s** page issue was more annoying than anticipated. In the [last commit](https://github.com/sourcegraph/sourcegraph/pull/59639/commits/6f2e6f76a8ab1e461756dde9a9d3ed7e29a03a93) I changed the page back to an infinity scroll list. Arguably my way of making paginated back navigation work was very hacky and didn't work well with using Apollo for caching the data.
But that alone wasn't enough. The whole approach with using Apollo for pagination caching only seems to be easy if there is only one place that uses the data in an infinity scroll list.
The main reason was that both the history panel and the commits page use the same cached data, but the history panel requests fewer fields. So if you open the commits after opening the history panel on the repo root page, new requests would be made to fetch the missing fields, which would mess up the simple concatenation cache. Trying to merge income and existing data didn't work either.

So in the end I resorted to ensuring that both queries fetch the same data (left comments in the gql files). I hope we'll revamp the commits at some point and arrive at a cleaner solution. Or maybe we do have to go back to manually managing the loaded data state in the component.
